src/qz/utils/FileWatcher.java

```diff
@@ -0,0 +1,132 @@
+                    alive = false;
+                }
+            }
+        });
+        watchThread.start();
+    }
+
+    public static void registerWatch(FileIO fileIO) throws IOException {
+        fileIO.setWk(fileIO.getAbsolutePath().register(watchService,
+                                                       StandardWatchEventKinds.ENTRY_MODIFY,
+                                                       StandardWatchEventKinds.ENTRY_CREATE,
+                                                       StandardWatchEventKinds.ENTRY_DELETE));
+
+        fileIOs.add(fileIO);
+    }
+
+    public static void deregisterWatch(FileIO fileIO) {
+        fileIOs.remove(fileIO);
+    }
+
+
+    private synchronized static void fileChanged(Path path, String fileName, String type) {
+        Path filePath = path.resolve(fileName);
+        for(FileIO fio : fileIOs) {
+            String fileData = null;
+            if (fio.getAbsolutePath().equals(path.normalize().toAbsolutePath())) {
+                if (!type.equals("ENTRY_DELETE") && fio.returnsContents() && !Files.isDirectory(filePath)) {
+                    try {
+                        switch(fio.getReadType()) {
+                            case BYTES:
+                                fileData = getBytes(filePath, fio);
+                                break;
+                            case LINES:
+                            default:
+                                fileData = getLines(filePath, fio);
+                        }
+                    }
+                    catch(IOException e) {
+                        log.error("Failed to read file due to {}", e.toString());
+                        fio.sendError("Failed to read file data due to " + e.getClass().getName());
+                    }
+                }
+
+                fio.fileChanged(fileName, type, fileData);
+            }
+        }
+
+    }
+
+    private synchronized static String getBytes(Path path, FileIO listener) throws IOException {
+        try(RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r")) {
+            byte[] bytes = listener.getBytes() == -1? new byte[(int)raf.length()]:new byte[(int)Math.min(raf.length(), listener.getBytes())];
+            if (listener.isReversed()) { raf.seek(raf.length() - bytes.length); }
+            raf.readFully(bytes);
+
+            return new String(bytes, Charset.forName("UTF-8"));
+        }
+    }
+
+    private synchronized static String getLines(Path path, FileIO listener) throws IOException {
+        StringBuilder sb = new StringBuilder();
+        String buffer;
+
+        if (listener.isReversed()) {
+            try(ReversedLinesFileReader reader = new ReversedLinesFileReader(path.toFile())) {
+                int count = 0;
+                while((buffer = reader.readLine()) != null && count++ != listener.getLines()) {
+                    // Warning, this will strip "\r" from the data
+                    sb.append(buffer).append("\n");
```

Comment:
After testing out `reversed`, I can't see a logical or intuitive reason to return the lines in reverse order.  Thoughts welcome, but I think if you're asking for the last 10 lines of a file, they should still be in the natural order.

For example:

First 10:
```
test1
test2
test3
test4
test5
test6
test7
test8
test9
test10
```

Last 10:
```
test7
test8
test9
test10
test11
test12
test13
test14
test15
test16
```